### PR TITLE
Reconcile Extensions Secret with EEC Token

### PR DESCRIPTION
**Pull request by albertogdd**
Follow-up from https://github.com/Dynatrace/dynatrace-operator/pull/3426

## Description

[K8S-10120](https://dt-rnd.atlassian.net/browse/K8S-10120)

New extensions secret is reconciled when `dynakube.spec.extensions.prometheus.enabled = true`
Secret name = `<dk-name>-extensions-token`, containing `eec-token` data, which gets generated only once.

New conditions added related to the secret creation

## How can this be tested?

- Apply dynakube with `dynakube.spec.extensions.prometheus.enabled = true`
- Check that secret gets created
- Check that condition is added
- Check that condition gets deleted if prometheus gets disabled and the secret gets deleted

### CODE REVIEW

File: pkg/controllers/dynakube/extension/secret.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	eecTokenKey         = "eec-token"
+	eecTokenValuePrefix = "EEC dt0x01"
+	secretSuffix        = "-extensions-token"
+)
+
+func reconcileSecret(ctx context.Context, dk *dynakube.DynaKube, kubeClient client.Client, apiReader client.Reader) error {
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/secret.go:20]: let's do same patter as we do for dynatrace `func (r *reconciler) reconcileSecret(ctx context.Context) error {`

File: pkg/controllers/dynakube/extension/secret.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	log.Info("reconciling secret" + getSecretName(dk.Name))
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/secret.go:21]: may be we need add a space here ^

File: pkg/controllers/dynakube/extension/secret.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+	_, err := query.Get(client.ObjectKey{Name: getSecretName(dk.Name), Namespace: dk.Namespace})
+	if err != nil && !errors.IsNotFound(err) {
+		setSecretCreatedFailureCondition(dk.Conditions(), err)
+
+		return err
+	}
+
+	if errors.IsNotFound(err) {
+		log.Info("creating secret" + getSecretName(dk.Name))
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/secret.go:44]: space is missing

File: pkg/controllers/dynakube/extension/secret.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/Dynatrace/dynatrace-operator/pkg/api/v1beta3/dynakube"
+	"github.com/Dynatrace/dynatrace-operator/pkg/util/dttoken"
+	k8ssecret "github.com/Dynatrace/dynatrace-operator/pkg/util/kubeobjects/secret"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	eecTokenKey         = "eec-token"
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/secret.go:15]: in my PR i've created consts.go similar to other reconcilers, please add also this file and move it there.

File: pkg/controllers/dynakube/extension/secret.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	setSecretCreatedSuccessCondition(dk.Conditions())
+
+	return nil
+}
+
+func buildSecret(dk *dynakube.DynaKube, token dttoken.Token) (*corev1.Secret, error) {
+	secretData := map[string][]byte{
+		eecTokenKey: []byte(token.String()),
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/secret.go:75]:
```suggestion
		eecTokenSecretKey: []byte(token.String()),
```

File: pkg/controllers/dynakube/extension/secret.go (hunk 1 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/secret.go]: [Fixed here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/24ace1ad5534ddee99b63515d3e7b766683561bc)

File: pkg/controllers/dynakube/extension/secret.go (hunk 2 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/secret.go]: Fixed [here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/24ace1ad5534ddee99b63515d3e7b766683561bc)

File: pkg/controllers/dynakube/extension/secret.go (hunk 3 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/secret.go]: Fixed [here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/24ace1ad5534ddee99b63515d3e7b766683561bc)

File: pkg/controllers/dynakube/extension/secret.go (hunk 4 above)
Comment (andriisoldatenko) [reply pkg/controllers/dynakube/extension/secret.go:15]: https://github.com/Dynatrace/dynatrace-operator/pull/3487/files#diff-ffc8830719c7c7221619364b5ad1ee791546f72852e019535f4ff8b59c44f4a0

File: pkg/controllers/dynakube/extension/secret.go (hunk 5 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/secret.go]: [Fixed here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/ba5a535b2e26209d7c415d79cdd93824001c3f68)

File: pkg/controllers/dynakube/extension/secret.go (hunk 4 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/secret.go]: [Fixed here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/ba5a535b2e26209d7c415d79cdd93824001c3f68)

File: pkg/controllers/dynakube/extension/reconciler_test.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	testName      = "test-name"
+	testNamespace = "test-namespace"
+)
+
+func TestReconciler_Reconcile(t *testing.T) {
+	t.Run(`Prometheus disabled`, func(t *testing.T) {
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/reconciler_test.go:25]:
```suggestion
	t.Run(`Extension token not generated when Prometheus is disabled`, func(t *testing.T) {
```

File: pkg/controllers/dynakube/extension/reconciler_test.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+		// assert extensions token is not generated
+		var secretFound corev1.Secret
+		err = fakeClient.Get(context.Background(), client.ObjectKey{Name: testName + "-extensions-token", Namespace: testNamespace}, &secretFound)
+		require.True(t, k8serrors.IsNotFound(err))
+
+		// assert conditions are empty
+		require.Empty(t, instance.Conditions())
+	})
+	t.Run(`Prometheus enabled`, func(t *testing.T) {
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/reconciler_test.go:42]: same here

File: pkg/controllers/dynakube/extension/reconciler_test.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+		require.NoError(t, err)
+
+		// assert extensions token condition is added
+		require.NotEmpty(t, instance.Conditions())
+		var expectedConditions []metav1.Condition
+		setSecretCreatedSuccessCondition(&expectedConditions)
+		testutil.PartialEqual(t, &expectedConditions, instance.Conditions(), cmpopts.IgnoreFields(metav1.Condition{}, "LastTransitionTime"))
+	})
+
+	t.Run(`Secret creation failed`, func(t *testing.T) {
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/reconciler_test.go:62]: and here 🙏 🙏 🙏

File: pkg/controllers/dynakube/extension/reconciler_test.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+		// assert extensions token condition is added
+		require.NotEmpty(t, instance.Conditions())
+		var expectedConditions []metav1.Condition
+		setSecretCreatedFailureCondition(&expectedConditions, err)
+		testutil.PartialEqual(t, &expectedConditions, instance.Conditions(), cmpopts.IgnoreFields(metav1.Condition{}, "LastTransitionTime"))
+	})
+}
+
+func makeTestDynakube() *dynatracev1beta3.DynaKube {
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/reconciler_test.go:78]:
```suggestion
func makeTestDynakube() *dynakube.DynaKube {
```

File: pkg/controllers/dynakube/extension/reconciler_test.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
+package extension
+
+import (
+	"context"
+	"testing"
+
+	"github.com/Dynatrace/dynatrace-operator/pkg/api/scheme/fake"
+	dynatracev1beta3 "github.com/Dynatrace/dynatrace-operator/pkg/api/v1beta3/dynakube"
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/reconciler_test.go:8]:
```suggestion
	"github.com/Dynatrace/dynatrace-operator/pkg/api/v1beta3/dynakube"
```
Please read new section I've added recently https://github.com/Dynatrace/dynatrace-operator/blob/main/doc/coding-style-guide.md#naming

File: pkg/controllers/dynakube/extension/conditions.go [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+	}
+	_ = meta.SetStatusCondition(conditions, condition)
+}
+
+func setSecretCreatedFailureCondition(conditions *[]metav1.Condition, err error) {
+	condition := metav1.Condition{
+		Type:    secretConditionType,
+		Status:  metav1.ConditionFalse,
+		Reason:  secretCreatedReason,
+		Message: fmt.Sprintf(secretCreatedMessageFailure, err),
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/conditions.go:25]: just a nitpick, but i think it's not a good idea to use constant here, and better to put content directly:
```suggestion
		Message: "Error creating extensions secret: " + err.Error(),
```

similar to https://github.com/dynatrace/dynatrace-operator/blob/2a85bd2f4aded19b83f259c00fc6f645b4b6e1cd/pkg/util/conditions/error.go#L16-L28


or may be reuse it? not sure we want very specific condition for this case
cc @StefanHauth  @0sewa0 wdyt?

File: pkg/controllers/dynakube/extension/conditions.go [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func setSecretCreatedSuccessCondition(conditions *[]metav1.Condition) {
+	condition := metav1.Condition{
+		Type:    secretConditionType,
+		Status:  metav1.ConditionTrue,
+		Reason:  secretCreatedReason,
+		Message: secretCreatedMessageSuccess,
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/conditions.go:15]: same here, or move it to top of this file like in https://github.com/dynatrace/dynatrace-operator/blob/2a85bd2f4aded19b83f259c00fc6f645b4b6e1cd/pkg/controllers/dynakube/injection/conditions.go#L17

File: pkg/controllers/dynakube/extension/conditions.go (hunk 11 above)
Comment (0sewa0) [reply pkg/controllers/dynakube/extension/conditions.go:25]: https://github.com/Dynatrace/dynatrace-operator/blob/d03faddf20f699667d441657547e31ecd2cb9746/pkg/util/conditions/secret.go#L13-L43

File: pkg/controllers/dynakube/extension/reconciler_test.go (hunk 6 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/reconciler_test.go]: [Fixed here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/f288a82fe537eb2e859f3dca4fd369fafcf96094)

File: pkg/controllers/dynakube/extension/reconciler_test.go (hunk 7 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/reconciler_test.go]: [Fixed here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/f288a82fe537eb2e859f3dca4fd369fafcf96094)

File: pkg/controllers/dynakube/extension/reconciler_test.go (hunk 8 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/reconciler_test.go]: [Fixed here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/f288a82fe537eb2e859f3dca4fd369fafcf96094)

File: pkg/controllers/dynakube/extension/reconciler_test.go (hunk 9 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/reconciler_test.go]: [Fixed here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/f288a82fe537eb2e859f3dca4fd369fafcf96094)

File: pkg/controllers/dynakube/extension/reconciler_test.go (hunk 10 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/reconciler_test.go]: [Fixed here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/f288a82fe537eb2e859f3dca4fd369fafcf96094)

File: pkg/controllers/dynakube/extension/conditions.go (hunk 11 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/conditions.go]: Thanks for the input guys! I refactored it a bit and now I use the util.conditions package, [see here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/f288a82fe537eb2e859f3dca4fd369fafcf96094)

File: pkg/controllers/dynakube/extension/reconciler_test.go [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+		// assert extensions token is not generated
+		var secretFound corev1.Secret
+		err = fakeClient.Get(context.Background(), client.ObjectKey{Name: testName + "-extensions-token", Namespace: testNamespace}, &secretFound)
+		require.True(t, k8serrors.IsNotFound(err))
+
+		// assert conditions are empty
+		require.Empty(t, instance.Conditions())
+	})
+	t.Run(`Extension secret is generated when Prometheus is disabled`, func(t *testing.T) {
+		instance := makeTestDynakube()
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/reconciler_test.go:45]: test says: Prometheus is disabled but `makeTestDynakube` hardcode
```
Prometheus: dynakube.PrometheusSpec{
					Enabled: true,
				},
```

File: pkg/controllers/dynakube/extension/reconciler_test.go [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+func makeTestDynakube() *dynakube.DynaKube {
+	return &dynakube.DynaKube{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: testNamespace,
+			Name:      testName,
+		},
+		Spec: dynakube.DynaKubeSpec{
+			Extensions: dynakube.ExtensionsSpec{
+				Prometheus: dynakube.PrometheusSpec{
+					Enabled: true,
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/reconciler_test.go:89]: better to not hardcode it here, but change in tests explicitly.

File: pkg/controllers/dynakube/extension/reconciler_test.go (hunk 13 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/reconciler_test.go]: True! it was a typo. [Fixed](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/3def3f6a0eeb9547c6b6538281f4e8ce1e6569a9)

File: pkg/controllers/dynakube/extension/reconciler_test.go (hunk 14 above)
Comment (albertogdd) [reply pkg/controllers/dynakube/extension/reconciler_test.go]: Makes sense. [Refactored here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/53fab361accb85dff2627db62f248cbdd3323266)

File: pkg/controllers/dynakube/extension/reconciler_test.go [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+const (
+	testName      = "test-name"
+	testNamespace = "test-namespace"
+)
+
+func TestReconciler_Reconcile(t *testing.T) {
+	t.Run(`Extension secret not generated when Prometheus is disabled`, func(t *testing.T) {
+		instance := makeTestDynakube(false)
+		instance.Spec.Extensions.Prometheus.Enabled = false
```
Comment (andriisoldatenko) [inline pkg/controllers/dynakube/extension/reconciler_test.go:29]:
```suggestion
```

**Review by andriisoldatenko — DISMISSED** [review]
practically LGTM, tested in my cluster.

**Review by andriisoldatenko — APPROVED** [review]

**Review by aorcholski — COMMENTED** [review]
the operator reports many times "removing the EEC secret" if prometheus is disabled 
```
{"level":"info","ts":"2024-07-17T13:29:57.572Z","logger":"extension","msg":"start reconciling extensions"}
{"level":"info","ts":"2024-07-17T13:29:57.572Z","logger":"extension","msg":"reconciling secret dynakube-extensions-token"}
{"level":"info","ts":"2024-07-17T13:29:57.572Z","logger":"extension","msg":"removing secret","name":"dynakube-extensions-token","namespace":"dynatrace"}

{"level":"info","ts":"2024-07-17T13:30:21.355Z","logger":"extension","msg":"start reconciling extensions"}
{"level":"info","ts":"2024-07-17T13:30:21.355Z","logger":"extension","msg":"reconciling secret dynakube-extensions-token"}
{"level":"info","ts":"2024-07-17T13:30:21.355Z","logger":"extension","msg":"removing secret","name":"dynakube-extensions-token","namespace":"dynatrace"}
```

**Comment by albertogdd** [comment]
> the operator reports many times "removing the EEC secret" if prometheus is disabled
> 
> ```
> {"level":"info","ts":"2024-07-17T13:29:57.572Z","logger":"extension","msg":"start reconciling extensions"}
> {"level":"info","ts":"2024-07-17T13:29:57.572Z","logger":"extension","msg":"reconciling secret dynakube-extensions-token"}
> {"level":"info","ts":"2024-07-17T13:29:57.572Z","logger":"extension","msg":"removing secret","name":"dynakube-extensions-token","namespace":"dynatrace"}
> 
> {"level":"info","ts":"2024-07-17T13:30:21.355Z","logger":"extension","msg":"start reconciling extensions"}
> {"level":"info","ts":"2024-07-17T13:30:21.355Z","logger":"extension","msg":"reconciling secret dynakube-extensions-token"}
> {"level":"info","ts":"2024-07-17T13:30:21.355Z","logger":"extension","msg":"removing secret","name":"dynakube-extensions-token","namespace":"dynatrace"}
> ```

[Fixed here](https://github.com/Dynatrace/dynatrace-operator/pull/3485/commits/82f558126ae7c1c7067f9c365669eaf84855359a)

**Review by aorcholski — APPROVED** [review]
LGTM

**Review by andriisoldatenko — APPROVED** [review]

**Pull request merged**